eclipsesource/theia
# Reapply changes for new message rpc

#### What it does

Reapplies the changes for the binary message rpc, originally introduced with #11011 and fixes the critical issues with lead to reverting the original change. 

In particular the following issues are addressed:

- #11199  Ensure that the new message rpc properly handles backend disconnects/reconnects.
- #11203 Ensure that debugging is handled gracefully (implicitly fixed with the fix for #11199)

These bugs were caused by a combination of  underlying issues:
- Trying to send RPC calls on a already disposed/disconnected websocket channel resulted in an endless loop due to using `console.warn`. Using console logging in the browser indirectly also triggers an RPC call to the `LoggerServer`. Naturally this call also fails and due to the console logging triggers another rpc call => we end up in and endless loop. Fixed by removing the console.warn statement in `websocket-channel.ts`.
- On socket connection loss (i.e. backend disconnect) channels and `RpcProtocol` instances were only partly disposed. This means once the backend reconnected did end up with multiple `RpcProtocl` instances for the same proxy which lead to weird issues e.g. Sending a request from one instance but the reply for the request is then handled by the other instance.
This is now fixed by ensuring that all components involved in the rpc communication (websocket main channel, multiplexer, multiplexed service channels, `RpcProtocol` instances) are fully disposed on disconnect and re-instantiated on reconnect.
- On reconnect the `connection-service` was informed too early i.e. before all previously open channels have been restored. As a consequence some RPC calls (e.g. ping requests) were triggered before the responsible service channel was available.
Fixed in `WebSocketConnectionProvider` by ensuring that the `onSocketDidOpen` event is fired after all previously openend channels have been restored.
- The `onIncomingMessgeActivity` event of the `WebSocketConnectionProvider` was never fired.
### How to test

To test the fix for #111999:

- Start the browser application and connect in a browser.
- Kill the backend.
- Application status should change color to indicate the lost connection.
- Restart backend
- Application should reconnect and work as before the connection loss

To test the fix for #11203 
- In the browser application (may also apply to Electron)
- Set a break point in the developer tools and hit it.
- Wait a few minutes.
- Remove your breakpoint and hit continue.
- Debugging process should continue without any issues and application should be responsive. No errors about failures to send messages should be logged.



#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
